### PR TITLE
Python implementation of the message filter TimeSequencer

### PR DESCRIPTION
**Pull request by spaghetti-**
https://github.com/ros/ros_comm/issues/564

Tested with https://gist.github.com/spaghetti-/6e11c0789363a66456bc however I do agree that this isnt a very comprehensive test. 

I would love verification/testing from anyone.

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Please make the new code follow PEP 8 / 257 and also include documentation for the constructor arguments.

File: utilities/message_filters/src/message_filters/__init__.py [hunk 1]
```diff
@@ -226,3 +226,46 @@ def add(self, msg, my_queue):
...
+
+    def __init__(self, fs, delay, rate, queue_size):
+        SimpleFilter.__init__(self)
+        self.lock       = threading.Lock()
+        self.queue      = {}
+        self.queue_size = queue_size
+        self.last_stamp = rospy.Time(0)
+        self.rate       = rospy.Duration.from_sec(rate)
+        self.delay      = rospy.Duration.from_sec(delay)
+        self.timer      = rospy.Timer(self.rate, self.check)
```
Comment (dirk-thomas) [inline utilities/message_filters/src/message_filters/__init__.py:252]: Please don't align the equal signs.

File: utilities/message_filters/src/message_filters/__init__.py [hunk 2]
```diff
@@ -226,3 +226,46 @@ def add(self, msg, my_queue):
...
+        fs.registerCallback(self.add)
+
+    def add(self, msg):
+        self.lock.acquire()
+        if msg.header.stamp > self.last_stamp:
+            self.queue[msg.header.stamp] = msg
```
Comment (dirk-thomas) [inline utilities/message_filters/src/message_filters/__init__.py:258]: This seems to overwrite other messages with the same timestamp?

File: utilities/message_filters/src/message_filters/__init__.py [hunk 3]
```diff
@@ -226,3 +226,46 @@ def add(self, msg, my_queue):
...
+        while len(self.queue) > self.queue_size:
+            t = min(self.queue)
+            del self.queue[t]
+        self.lock.release()
```
Comment (dirk-thomas) [inline utilities/message_filters/src/message_filters/__init__.py:262]: Please use `with` instead to ensure unlock even when exceptions are raised.

Same below.

**Comment by spaghetti-** [comment]
Thanks for your comments, will update the issues and test against the C++ TimeSequencer as well.

However when I ran pep8 against the file there are many other issues that are raised, should I address them in a different PR and just fix for my changes here?

**Comment by dirk-thomas** [comment]
Just make sure that the newly added code block is "clean". You should not update existing code since it makes backporting patches way more difficult. Thanks.

File: utilities/message_filters/src/message_filters/__init__.py [hunk 4]
```diff
@@ -226,3 +226,46 @@ def add(self, msg, my_queue):
...
+
+    def check(self, _):
+        self.lock.acquire()
+        to_call = sorted(filter(lambda t: t + self.delay <= rospy.Time.now(),
+                              self.queue.keys()))
```
Comment (mikepurvis) [inline utilities/message_filters/src/message_filters/__init__.py:267]: I suppose it depends on the length of the queue and frequency of this function being called, but sorting the dictionary keys every time seems not ideal.

I wonder what the performance difference would be using a deque which stays sorted by inserting new messages into the correct position.

**Comment by dirk-thomas** [comment]
Closing due to long time of inactivity and targeting an EOL branch. Please consider opening a new PR targeting the current default branch if you are still interested in getting this patch merged.

**Pull request closed without merging**